source/validate_logicals.cpp

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Ensures type declarations are unique unless allowed by the specification.
+
+#include "validate.h"
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+namespace {
+
+// Returns operand word for given instruction and operand index.
+// The operand is expected to only have one word.
+inline uint32_t GetOperandWord(const spv_parsed_instruction_t* inst,
```

Comment:
All calls to `GetOperandWord` are prepended with `_.GetTypeId`. For example:
```cpp
const uint32_t operand_type = _.GetTypeId(GetOperandWord(inst, index));
```

I suggest you change the function to:
```cpp
inline uint32_t GetOperandType(
  ValidationState_t& _ ,
  const spv_parsed_instruction_t * inst,
  size_t operand_index)
```
and the change the function to return the type:
```cpp
return _.GetTypeId(inst->words[operand.offset]);
```
This will simplify the call sites

Reply:
This is a good simplification of the idiom.  But I suggest appending "Id" to the name, to get GetOperandTypeId  because we might want to save GetOperandType so it returns an ir::Instruction pointer.